### PR TITLE
An (almost) empty GeoMechanicsApplication to be merged to master

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**Description**
As the size of GeoMechanicsApplication is very large, we decided to split it into smaller PRs. This is the minimum application that I created for GeoMechanicsApplication as a starting point.

The featuers of this will be added gradually.

**Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- add minimum GeoMechanicsApplication

### CODE REVIEW

**Comment by philbucher** [comment]
looks good
I have some comments that I will add in the next days

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py]:
```suggestion
```

no longer needed as we dropped support for python2

File: applications/GeoMechanicsApplication/geo_mechanics_application.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application.h:29]:
```suggestion
```
this should really not be here

File: applications/GeoMechanicsApplication/README.md
Comment (philbucher) [inline applications/GeoMechanicsApplication/README.md:50]: this is added automatically though `kratos_add_dependency` in your cmake

File: applications/GeoMechanicsApplication/README.md
Comment (philbucher) [inline applications/GeoMechanicsApplication/README.md:49]: you sure you want to add this?
Technically it is not necessary

maybe make a comment that it is recommended?

File: applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp:46]: you don't want to expose your variables?

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application.cpp:138]: we have macros for tensor variables

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application.cpp]: hm those names will be very hard to understand and debug
what is the intended usecase?
Maybe there is a better solution

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:60]:
```suggestion
    enum class indexStress2DPlaneStrain{ INDEX_2D_PLANE_STRAIN_XX,
```
more C++11

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:41]: what is the intended usecase for those?

File: applications/GeoMechanicsApplication/geo_mechanics_application_variables.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application_variables.h]: those are heavy includes

consider including them only where they are actually used

File: applications/GeoMechanicsApplication/geo_mechanics_application_variables.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application_variables.h:22]:
```suggestion
```

this one should not be here

**Review by philbucher — COMMENTED** [review]
some initial comments

I have some more for the solvers but no time now, will add them in the next days

**Review by RiccardoRossi — COMMENTED** [review]
i subscribe the comments of @philbucher 
other than that ok to me

File: applications/GeoMechanicsApplication/geo_mechanics_application.h
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application.h:29]: Resolved.

File: applications/GeoMechanicsApplication/README.md
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/README.md:50]: I will update README.md later.

File: applications/GeoMechanicsApplication/README.md
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/README.md:49]: I will update README.md at a later stage.

File: applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/custom_python/geo_mechanics_python_application.cpp:46]: Resolved.

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application.cpp]: PARAMETER_xxx are replaced with a vector. The state_varaible_xxx are needed only for output, if a user wants to retrieve a specific state variables. The state variables in UMATs do not have names.

File: applications/GeoMechanicsApplication/geo_mechanics_application_variables.h
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application_variables.h]: reduced

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:41]: I want to use compiler independent limits for large and small numbers.
The pore_pressure_sign_factor is used to simply switch between soil mechanics and structural mechanics sign convention for water pressure (stress). By default,

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:60]: I tried but this creates a lot of errors. I leave it as it is for the time being.

**Comment by Vahid-Galavi** [comment]
I think, I addressed all the issues. Could you please check if you can approve this PR in order to add the application?

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py]: this was removed in Structural
it was a huge pain, I strongly recommend doing the same

**Review by philbucher — COMMENTED** [review]
@Vahid-Galavi it seems like the python solvers are largely copy pasted from structural and have LOTS of outdated stuff.
=> why don't you derive your base-solver from the base structural solver?

since those are sensitive and important things I recommend making a separate PR for them
or not up to you, I gave my comments :)

File: applications/GeoMechanicsApplication/geo_mechanics_application.cpp
Comment (RiccardoRossi) [reply applications/GeoMechanicsApplication/geo_mechanics_application.cpp:138]: @loumalouomega can u point to one example of definition of those matrix variables and components?

NOTE of @loumalouomega, here are the macros:
https://github.com/KratosMultiphysics/Kratos/blob/53f15798b4dc46d8a0f60c3ca1cbece0be3a1557/kratos/includes/define.h#L199

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py]: this could also be removed

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
from __future__ import absolute_import, division #makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```

can be removed

**Review by RiccardoRossi — COMMENTED** [review]
@Vahid-Galavi i was looking into this PR, and then i realized that you are modifying the big PR instead of this one, i guess to then cherry pick the changes?

If you think it is more confortable, we can give you the ok to merge this one so to simplify the procedure. The only thing that we see as important is not to have 500 unreviewed files to go in at once.

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py]: removed!

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]: removed

File: applications/GeoMechanicsApplication/geo_mechanics_application_variables.h
Comment (philbucher) [inline applications/GeoMechanicsApplication/geo_mechanics_application_variables.h:98]: this was removed in structural

=> now we use `COMPUTE_LUMPED_MASS_MATRIX` from the Core

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
```

this is super old

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
```

those should only be imported where needed

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
        super().__init__(model,parameters)
```

python 3 syntax

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]: imports should be at the top of the file

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
```

comments should be removed

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
```

this is also super old and was removed a long time ago in Structural

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py]:
```suggestion
        super().FinalizeSolutionStep()
```

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py]:
```suggestion
        super().__init__(model, custom_settings)
```

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py]:
```suggestion
```

no longer necessary

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py]:
```suggestion
```

was removed long ago in Strcutral

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py
Comment (philbucher) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_solvers_wrapper.py]:
```suggestion
```

**Comment by Vahid-Galavi** [comment]
> @Vahid-Galavi i was looking into this PR, and then i realized that you are modifying the big PR instead of this one, i guess to then cherry pick the changes?
> 
> If you think it is more confortable, we can give you the ok to merge this one so to simplify the procedure. The only thing that we see as important is not to have 500 unreviewed files to go in at once.

I think, we should speed up the procedure, otherwise this application won't be merged. Even merging an almost empty application that I started 5 days ago has not been done.

**Comment by RiccardoRossi** [comment]
@Vahid-Galavi this is exactly what we are trying... 

you made an almost empty app and we posted comments to try to make it on the same line as the structural mechanics.
If you like we are willing to accept it as-is.
In our experience refactoring the solvers after merging it is very painful. This is why we think it's worth it to spend a bit more effort in this very first stage rather than in future steps. Also note that aligning the developments with the StructuralMechanicsApplication ones will make our lives easier as the maintenance will be somehow automated.

File: applications/GeoMechanicsApplication/geo_mechanics_application_variables.h
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/geo_mechanics_application_variables.h:98]: This part will definitely be removed. All the elements that I borrowed from Structural Mechanics will be removed as soon as the reseting displacement issue is solved in that application.

**Review by RiccardoRossi — DISMISSED** [review]
approving it as-is under the understanding that more cleanups will follow

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:1]: Just mentioning that I like this idea of having the required constants in a separated file rather than adding them in the application ones.

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py
Comment (rubenzorrilla) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py:65]:
```suggestion
            "convergence_criterion": "displacement_criterion",
```
Otherwise the convergence criteria factory would throw an error if not provided.

Comment (rubenzorrilla) [reply applications/GeoMechanicsApplication/python_scripts/geomechanics_U_Pw_solver.py:65]: OK. Now I see the function below. It works. Feel free to dismiss this comment.

**Review by rubenzorrilla — APPROVED** [review]
I think that in general terms this can go in.

However I've a minor comments would be nice to clarify and address if required (in this PR or in further ones). These are:
- The issue of the tensor variables (i.e. `PERMEABILITY`)
- The `documentation` folder. There is a lot of files in there, which might eventually become large (e.g. *.png and *.pdf). Considering that the documentation issue is being a hot topic in the community we can wait until we have a final decision to take any action. However, I think that we should move these to another place better suited for that.
- I wonder why you need two analysis stage classes (one deriving from the other). I guess it is something related to the multistage. As we are starting to sketch the final design of the multistage analysis, it would be really helpful for us if you can clarify this a bit.

Thanks for the patience :wink:

**Comment by Vahid-Galavi** [comment]
Mer

> I think that in general terms this can go in.
> 
> However I've a minor comments would be nice to clarify and address if required (in this PR or in further ones). These are:
> 
> * The issue of the tensor variables (i.e. `PERMEABILITY`)
> * The `documentation` folder. There is a lot of files in there, which might eventually become large (e.g. *.png and *.pdf). Considering that the documentation issue is being a hot topic in the community we can wait until we have a final decision to take any action. However, I think that we should move these to another place better suited for that.
> * I wonder why you need two analysis stage classes (one deriving from the other). I guess it is something related to the multistage. As we are starting to sketch the final design of the multistage analysis, it would be really helpful for us if you can clarify this a bit.
> 
> Thanks for the patience 😉

Thanks for reviewing.
- I couldn't find any example how a tensorial variable can be defined. This will reduces size of the code.
- The document folder is anyway needed for this application and it will grow. It would be nice to discuss later what to do with documentations.

**Comment by rubenzorrilla** [comment]
> Mer
> 
> > I think that in general terms this can go in.
> > However I've a minor comments would be nice to clarify and address if required (in this PR or in further ones). These are:
> > 
> > * The issue of the tensor variables (i.e. `PERMEABILITY`)
> > * The `documentation` folder. There is a lot of files in there, which might eventually become large (e.g. *.png and *.pdf). Considering that the documentation issue is being a hot topic in the community we can wait until we have a final decision to take any action. However, I think that we should move these to another place better suited for that.
> > * I wonder why you need two analysis stage classes (one deriving from the other). I guess it is something related to the multistage. As we are starting to sketch the final design of the multistage analysis, it would be really helpful for us if you can clarify this a bit.
> > 
> > Thanks for the patience wink
> 
> Thanks for reviewing.
> 
>     * I couldn't find any example how a tensorial variable can be defined. This will reduces size of the code.
> 

I guess that redefining this variable would require to modify the elements. I think it would be easier to do it in the PR adding these with their corresponding unittests.  The macro to register tensor variables is `KRATOS_REGISTER_SYMMETRIC_3D_TENSOR_VARIABLE_WITH_COMPONENTS`. In the `define.h` you can find the other macros needed. An example on how to use them can be found in the `MeshingApplication`.

>     * The document folder is anyway needed for this application and it will grow. It would be nice to discuss later what to do with documentations.

The documentation folder would be needed for everything :rofl:. You can follow the documentation discussion evolution in [here](https://github.com/KratosMultiphysics/Kratos/discussions/8377).

**Pull request merged**
